### ISSUE TITLE
Presenting VC without a VP (non key-bound credentials)

### ISSUE THREAD

**Issue by OIDF-automation** [openid/OpenID4VP#6]
### Original Reporter: KristinaYasuda

A use-case being… A doctor is employed by an agency B, who has a contract with hospital A. So to prove that he works for hospital A, the doctor needs to send a VC1 that he works for agency B hospital A and VC2 that the agency B is contracted by hospital A. 

Doctor can create a VP1 of VC1, because it is issued to him, but he cannot create a VP of VC2 because VC2 is issued to an agency. but it should be ok as long as verifier trusts hospital’s signature. 

I think theoretically, it could be expressed in a descriptor\_map in a presentation\_submission that one VC can be found inside a VP and another VC is not inside a VP but as-is?

**Comment by alenhorvat** [comment]
The issue is not related to VP and signing, but to data modelling and expressing the relationships.
The two VCs are different in nature.

Question is: does the agency has the right/mandate to state that the doctor employed by them is working at the hospital?

To express such relationships we use "verifiable accreditations" (one could call them mandates, ...). Accreditations are statements about what a given entity can attest (what VC it can issue). For us this scenario is a non-problem, since the credentials are designed in a way they reference the accreditation. The accreditation can be either
a) published in a registry
b) part of the signature (signer attribute JOSE header as per ETSI).

How the model applies to this scenario:
Hospital issues a VC to the agency stating: you can attest that people employed by you, work at our hospital. Issuer: hospital, subject == holder == agency

When the agency issues a VC to the doctor, it references (can be a local reference like hash) the VC and puts it in the signer header (signed JOSE header). This way the statement is part of the agency-created signature.

When a Doctor presents a VC, it simply includes only the VC that has been issued to her. Namely, the hospital issued VC will already be in the header.

VP is signed by the doctor
VP contains a single VC (VC issued to the doctor by the agency)
VC signature (created by the agency) contains the VC issued to them by the hospital.

Note that the accreditation/mandate/other statements required for verification, can be transported in the unsigned JOSE header or in a VP.

**Comment by bit-jniestroj** [comment]
> ### Imported from AB/Connect bitbucket - Original Commenter: tlodderstedt
> Daniel, Kristina and I had a conversation about this topic at Identiverse. We agreed it is important to decide whether we always assume a key binding of the presented credentials (i.e. there is always a signature as proof of possession, which also allows cryptographic binding of nonce and audience of the response) or whether we see the need and would support presentation of credentials without key binding. Daniel took the task to create a write up as basis for further discussions.

@tlodderstedt One year ago you mentioned, that Daniel wanted to prepare a basis for further discussions regarding the topic if also credentials without a key binding are supported. Are there any news on that topic?

**Comment by jogu** [comment]
We did have a related discussion here relatively recently about this: https://github.com/openid/OpenID4VP/issues/161 and also https://github.com/openid/OpenID4VP/issues/61

My belief is no particular protocol changes are needed for this, but that we do need to finesse some of the language in definitions (e.g. `vp_token` is constrained to be a "verifiable presentation" which is in turn constrained to be "holder bound credentials"; widening that definition seems sensible.) Does that match your understanding?

**Comment by bit-jniestroj** [comment]
Yes i agree with that. From a protocol level there should be no changes as vp_token could be literally anything. Its just the wording that currently "prevents" credentials without a key binding.

Great to hear that this is still ongoing. If we can support somewhere just ping me. We will have use cases where we don't have a key binding and it would be awesome if that would also comply to the wording of the specification

**Comment by paulbastian** [comment]
@Sakurann @tlodderstedt @jogu If this is scheduled for 1.1, it looks like a breaking change to me. Currently section 14.1 reads very strong, allowing only Credentials with cryptographic binding. I think there are legitimate use cases not having such a binding, e.g. using claim-based binding

**Comment by paulbastian** [comment]
This is also in contradiction to [OpenID4VCI](https://openid.github.io/OpenID4VCI/openid-4-verifiable-credential-issuance-wg-draft.html#section-13.2)?

**Comment by danielfett** [comment]
There are a couple of ways we could implement this feature:

### DCQL Request

Encoding that a credential doesn't need a presentation in the response should be fairly easy to add to DCQL:

```json
{
  "credentials": [
    {
      "id": "my_credential",
      "format": "dc+sd-jwt",
      "meta": {
        "vct_values": [ "https://credentials.example.com/identity_credential" ]
      },
      "no_binding_needed": true,
      "claims": [
          {"path": ["last_name"]},
          {"path": ["first_name"]},
          {"path": ["address", "street_address"]}
      ]
    }
  ]
}
```

### Response

**Solution A** - Enforce that at least one credential with key binding is presented in each response. Any number of credentials without key binding can be presented at the same time (as requested in the DCQL query).

```json
{
  "cred_with_kb": "eyJhbGci...QMA~eyJh...kEA",
  "other_cred": "eyJhbGci...QMA"
}
```

This way, the nonce would always be available for checking. We would keep practically all security guarantees for the response. Of course, the non-key-bound credential could be replayed, but that is inherent to any presentation without key binding.

The protocol would not be able to support cases where only non-key-bound credentials are presented.

**Solution B1** - If no key-bound credential is presented, send the nonce, audience and whatever else we think is important in the VP Token in the response as separate, unsigned items. 
A problem would be to ensure that there are no clashes between the nonce/aud and a credential. Squigglies to the rescue, as they're not allowed in IDs in DCQL.

```json
{
  "cred_with_kb": "eyJhbGci...QMA~eyJh...kEA",
  "~unsigned": {"nonce": "4hqnKvw5l9", "aud": "client.example"}
}
```


Another problem is that the nonce and audience are now not cryptographically protected. That might be okay, however, as injection of responses is still hard.

A small risk is that Wallets always send the nonce and audience like this, even if they're not needed, and verifiers start relying on them for comparing what's in the presentation for key-bound credentials (i.e., comparing the signed nonce with the plaintext nonce from the response without checking it against the session). The countermeasure would be to define that verifiers MUST NOT accept responses containing these values AND a signed presentation.

**Solution B2** - As above, but send nonce and audience in the response as separate parameters outside of the VP Token.

`vp_token={...}&aud=...&nonce=...&state=...`

**Comment by alenhorvat** [comment]
What about setting the alg=none in kb-jwt?

**Comment by danielfett** [comment]
> What about setting the alg=none in kb-jwt?

Not allowed in SD-JWT (for good reasons, it only led to problems in the past).

**Comment by tlodderstedt** [comment]
> What about setting the alg=none in kb-jwt?

I have a bad feeling to use alg none because using a object that seems to be signed and is not is being perceived as a bad idea in general. Additionally, that would only work for sd-jwt and jwt, but not for mdoc (and other credential formats).

**Comment by alenhorvat** [comment]
But why would the protocol need to know anything about whether the credential is bound or not?
Does the security of the protocol rely on whether or not presentation is signed?

If yes, then always require that the presentation (kb-jwt in this case) is signed.

Whether or not to check the key binding, depends on whether cnf is present or not (or equivalent claim in other formats)

**Comment by paulbastian** [comment]
1. if there is a "cnf"-> always send KB, if there is none-> don't send it
2. make it very uncomplicated, just don't use the nonce and send te SD-JWT VC as is without KB

**Comment by Sakurann** [comment]
WG discussion:
- what Paul put in the comment above ^
- do not do anything in DCQL request and response, and **only remove the requirement that credentials in openid4vp have to be key bound and add security considerations that are being lost when credential is not key bound**.
    - can also add a section "do not use this in HAIP"? some disagreements.
- there are some credential formats that are hard to tell if they are session bound. might need something that makes it harder to screw it up.
- suggestion: mandate encryption for non-key bound credentials? might be not necessary?
- nonce is always in mdoc sessionTranscript.
- Daniel is sure something needs to be added in the response: nonce in the option or mandate `state`. only injection and CSRF is prevented

**Comment by alenhorvat** [comment]
Is it necessary to integrity protect the disclosures when presented?

**Comment by alenhorvat** [comment]
Is non-repudiation required when presenting credentials or not? (regardless of whether the credentials are cryptographically bound or not)

**Comment by alenhorvat** [comment]
> This is not usually required when presenting non-key-bound credentials.

Proof of possession and non-repudiation are two distinct properties. Maybe to check with the use cases/LSPs?

**Comment by danielfett** [comment]
@alenhorvat Do you have a use case in mind where non-repudiation with a non-key-bound credential could be required? Generally, I see very few use cases requiring non-repudiation (mainly use cases that make use of transaction data).

**Comment by alenhorvat** [comment]
A typical example (at least today) is that you sign the document, the document contains your personal information (e.g., social security or VAT number) and so does the x509 public key certificate.
Even though there's no explicit cryptographic binding, you have a binding via other identifiers. (one could talk about "identifier binding", which can be cryptographically protected)

> Do you have a use case in mind where non-repudiation with a non-key-bound credential could be required?

Personally I don't know an I haven't thought of this until now. Maybe a question for Steffen or Nacho.

**Comment by steffenschwalm** [comment]
> [@alenhorvat](https://github.com/alenhorvat) Do you have a use case in mind where non-repudiation with a non-key-bound credential could be required? Generally, I see very few use cases requiring non-repudiation (mainly use cases that make use of transaction data).


@danielfett non repudiation is needed in any regulated industry to make evident the transaction the RP did. Examples for Germany and internationally: Whole Public Sector in Germany acc. to Art. 20 Grundgesetz §§ 24 ff. VwVfG, 80 VwGO. Banking: AML, MiCAR, PSD2, Aeroapce: EASA Part 21, Pharma/LifeScience/MedTech/IManufacturing: FDA Part 11, AO, HGB, BGB, AtomG, GDPR; HealthCare: SGB V; Critical Infrastructure: NIS2, Telko: TKG, General: DSA, Travel: Public Security, IATA; Evidential value: §§ 371 ff. ZPO

Details see: Weber et al.: Records management acc. ISO 15489. Berlin 2018

See my issue in SPRIND GitHub due to which you changed your architecture in Germany you already accepted: https://gitlab.opencode.de/bmi/eudi-wallet/eidas-2.0-architekturkonzept/-/issues/50

So long Story short: > 80% of UseCases need non-repudiation to ensure documentation requirements and burden of proof for RP against 3rd parties. If EUDI Wallet shall not be used only for access your Disko in Leipzig or the next SPRIND Party - non-repudiation is essential for EUDI Wallet. 

Happy to give  OIDD ing general and SPRIND in particular a well-grounded introduction as I define the relevant standards since 2006 @tlodderstedt

**Comment by danielfett** [comment]
@steffenschwalm I was not asking about non-repudiation in general. I was asking if there is a specific use case in which a non-key-bound credential is presented and non-repudiation is required for this specific presentation.

**Comment by steffenschwalm** [comment]
@danielfett I see very few use cases requiring non-repudiation (mainly use cases that make use of transaction data). 

All of the mentioned use cases apply for this requirements

**Comment by steffenschwalm** [comment]
@danielfett technically after short alignment it seems your question is "Hence there should be an option to have signed presentations even when credentials are not bound to public keys, but maybe some other identifier, such as VAT?" the answer is: yes but not as default but as choice. The identity binding of (Q)EAA is legally not required so IMO there`s no need to bind it technically in all cases

**Comment by paulbastian** [comment]
I don't think we do ourselves a benefit in leaving any options to the holder, it will create unnecessary complications that is an additional risk of things going wrong. If the credentials is key-bound, then always send a keybinding during presentation. If the credential is not keybound, then just send the credential as-is without making protocols more complicated.

**Comment by alenhorvat** [comment]
@paulbastian, yes, but there's a separate question about the "holder non-repudiation" (which I don't have an answer to, hence I'm asking): is holder non-repudiation required, optional, or not required? (independently of the proof of possession use case)

**Comment by danielfett** [comment]
> [@danielfett](https://github.com/danielfett) technically after short alignment it seems your question is "Hence there should be an option to have signed presentations even when credentials are not bound to public keys, but maybe some other identifier, such as VAT?" the answer is: yes but not as default but as choice. The identity binding of (Q)EAA is legally not required so IMO there`s no need to bind it technically in all cases

What would the presentation be signed with, if there is no key?

> I don't think we do ourselves a benefit in leaving any options to the holder, it will create unnecessary complications that is an additional risk of things going wrong. If the credentials is key-bound, then always send a keybinding during presentation. If the credential is not keybound, then just send the credential as-is without making protocols more complicated.

I see this slightly differently: I agree that there is no point in asking the user. But there may be use cases where, from a Verifier's point of view, a presentation without PoP is sufficient *even if the credential supports it, i.e., has a cnf claim*. Therefore, it should be up to the Verifier to request it or not. Example: A mobile driver's license:

- Verifier A, where the user is signing up for a carsharing app, is interested in knowing that *this exact user* has a valid mdl. The Verifier would require the PoP.
- Verifier B, a car rental company, is just interested in learning that for all drivers, a valid license exists. This is a very common use case today, where I just upload a photo of my driver's license usually. I can upload the photo of my wife's license, the rental company knows that it is not mine, and they're fine with that. Hence, in the wallet space, they would not request PoP in the presentation of the mdl, at least for the second allowed driver.

> Example:
> 
>     * my diploma is not bound to any public key, but it's digitally signed by the university
> 
>     * my signed diploma is leaked
> 
>     * an attacker obtains my signed diploma and tries to use it to apply for XYZ
> 
>     * XYZ will need to do some identity matching, of course (so it will eventually figure out that the diploma doesn't belong to the attacker)
>       Questions are:
> 
>     * Should the unauthorised presentation of the diploma be blocked at the protocol level?
> 
>     * Does XYZ need a proof (signed presentation) that the diploma has been submitted (illegally)?
> 
> 
> If answer is "no" to both (I hope there are no other cases), then, as you mention [@paulbastian](https://github.com/paulbastian), signed presentation (kb-jwt) is essentially only a PoP that needs to be signed only when cnf is present.

Thanks for the example, @alenhorvat - this helps. I think the answer to the first question should be up to the Verifier - see my example above. The Verifier might be fine with accepting the replayed diploma (as long as it is a signed, official document) and then the Verifier would not require a signed KB-JWT.

**Comment by alenhorvat** [comment]
@danielfett, is the conclusion then that:
- if the credential is key-bound, kb-jwt may or may not be present, depending on whether the verifier requires the proof or not?
- if the credential is not key-bound, kb-jwt should not be present? (in practice, we have today signing keys - advanced/qualified e-signing keys for document signing; that key may be used to sign the presentation, hence the question is: is the kb-jwt needed (holder repudiation yes/no)? (design is of course simpler, if answer is "no"))

**Comment by steffenschwalm** [comment]
@danielfett 

> What would the presentation be signed with, if there is no key?

As @alenhorvat mentioned "in practice, we have today signing keys - advanced/qualified e-signing keys for document signing" - as it shall be possible to do this with EUDIW acc. to law I see no issue

@paulbastian We should trust in responsible user as he is owner of the credential and responsible for each transaction anyway.

**Comment by danielfett** [comment]
> [@danielfett](https://github.com/danielfett), is the conclusion then that:
> 
>     * if the credential is key-bound, kb-jwt may or may not be present, depending on whether the verifier requires the proof or not?

Correct.

>     * if the credential is not key-bound, kb-jwt should not be present? (in practice, we have today signing keys - advanced/qualified e-signing keys for document signing; that key may be used to sign the presentation, hence the question is: is the kb-jwt needed (holder repudiation yes/no)? (design is of course simpler, if answer is "no"))

Correct, if the credential is not key-bound, the kb-jwt should not be present. (Also by definition, it cannot be signed by a different key.) If a key for document signing is present and to be used to sign a document, that would be a distinct use case outside of the scope of this issue.

**Comment by steffenschwalm** [comment]
@danielfett @c2bo as EUDIW has to be able to sign independently from the issue here,, the goal is achieved with your suggestion. @alenhorvat do you agree`?

**Comment by Sakurann** [comment]
WG discussion

- no need to add anything with DC API, because API already ensures session binging
- but for non-DC API use-cases there is a need to bind it to the session (same reason as a `state`, but `state` has negative connotations?)

suggestion: 
if only a non-key bound credential is sent, send `state` alongside a VP Token (be careful with the language). and if non-key bound credential is sent alongside key bound credential, no need to send state. with a very strong recommendation ifnon-key bound credential is sent, there must be at least one key bound credentials alongside it.

**Comment by alenhorvat** [comment]
> Correct, if the credential is not key-bound, the kb-jwt should not be present. (Also by definition, it cannot be signed by a different key.) If a key for document signing is present and to be used to sign a document, that would be a distinct use case outside of the scope of this issue.

I agree, but my question is different and it seems nobody has an answer: is there a need by use cases to have "holder presentation non-repudiation" regardless of the key binding or not?

**Comment by danielfett** [comment]
Note: There is an overlap between the proposed DCQL parameter and the credential format idenfiers. For example, we currently distinguish `ldp_vc` and `ldp_vp` as format identifiers. That should essentially be sufficient for signalling "I need key binding" vs "I don't". But if we go that way, we would need to rename the SD-JWT format identifier, as it currently is `dc+sd_jwt` where it then should be `dc+sd-jwt-kb` or something (I just made the latter up).

**Comment by alenhorvat** [comment]
Some unstructured thoughts about the topic:

Is key binding something that needs to be expressed at the format level? As you mentioned: RP may or may not require a KB proof in the case cnf is present. (the driver's licence upload example)

Would then, the non-repudiation fall in the same category? But with non-repudiation it's slightly more complicated (hence the question, whether the presentation non-repudiation is needed or not, to which I don't have an answer);
RP could argue: regardless of the KB (which is a different property than non-repudiation) I need you to sign the presentation and the signature should contain your PII information so I can prove to an auditor you sent the credential (whether key binding is there or not).

As this complicates the discussion, maybe the approach to unblock this issue would be to cover the case of the key binding only (RP just signals whether it needs the KB proof or not); if KB is required and present, KB is generated and sent, if KB is not required and present or not present, KB is not created; if KB is required and not present, the credential does not qualify and no credential is shared.

The non-repudiation question can then be addressed separately, as it's slightly more elaborate.
